### ISSUE TITLE
Re-enable parallel gradle builds

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/fenix#13665]
Parallel gradle builds can greatly speed up build times, especially from a clean build. They were originally enabled in #1291 but disabled, without much of an explanation in, https://bugzilla.mozilla.org/show_bug.cgi?id=1512631 and its PR https://github.com/mozilla-mobile/android-components/pull/1588. In the `gradle.properties` file, the comment says, "# XXX Deactivated because of broken releases?"

If parallel gradle builds break releases in predictable way that we can ignore locally, **we should at least figure out how to enable it for developers locally.** Note: developers can work around this from the command line by running, `./gradlew <task> --parallel`

Running this on a clean build locally, I notice that `:samples-glean-library:Boostrap_CONDA_'Miniconda3'` fails. However, running `./gradlew clean :service-glean:Bootstrap_CONDA_\'Miniconda3\'` also fails so it appears the task may not be set up correctly. I filed #2344 for this and assume this bug is blocked on that.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-1789)

**Comment by data-sync-user** [comment]
➤ Michael Comella commented:

> ** Note: developers can work around this from the command line by running, ./gradlew <task> --parallel

Maybe CI should run a --noparallel, if it exists?

> Running this on a clean build locally, I notice that :samples-glean-library:Boostrap_CONDA_'Miniconda3' fails.

I wonder if parallel was disabled to begin with because this task is broken.

Once bootstrap runs successfully, e.g. run :service-glean:compiledebugkotlin, then parallel builds seems to work correctly: gr compiledebugkotlin --parallel.

**Comment by data-sync-user** [comment]
➤ Sebastian Kaspari commented:

--parallel caused multiple issues in automation - including intermittent test failures and a (perceived!) worse performance surprisingly.

**Comment by data-sync-user** [comment]
➤ Michael Comella commented:

Anecdotally, it seems to be significantly faster locally for clean builds (no tests) though I can see why it might be slower for incremental builds. I wonder if we could enable it for builds but disable it for tests? 🤔

**Comment by data-sync-user** [comment]
➤ Michael Comella commented:

Then again, people probably don't clean build often so it may not be worth the effort. Gradle 5+ might just fix this all anything. 😛

**Comment by data-sync-user** [comment]
➤ Michael Comella commented:

A few other benefits from this:

  * building faster on CI may decrease on CI spend
  * this may reduce friction for developing on combined fenix, a-c builds because fenix developers wouldn't have to wait as long for the initial build to use a-c (though there are other individual developer workarounds for this like automatically building master once a day just in case)

**Comment by data-sync-user** [comment]
➤ Richard Pappalardo commented:

@JohanLorenzo there have been a lot of updates this year including changeover to taskgraph.  is this worth re-trying?

**Comment by data-sync-user** [comment]
➤ Johan Lorenzo commented:

Good question! I don't think taskgraph itself can help solving this issue. That said, backporting https://github.com/mozilla-mobile/reference-browser/pull/1117 to a-c will likely help having faster and more reliable CI. Maybe --parallel can benefit from that.

**Comment by grigoryk** [comment]
Not sure why this was closed - we still don't have parallel builds enabled, neither locally nor in CI.

@mcomella do you know the state of this, anything new?

FWIW, I've just tried this locally on a clean Fenix build, and it worked just fine, while being almost 2m faster (5m56s vs 4m5s). ~30% speed up.

**Comment by grigoryk** [comment]
Enabling parallel execution in A-C also _appears_ to work locally, build times are cut in half (`publishToMavenLocal` went from ~2m to about ~1m).

**Comment by mcomella** [comment]
> @mcomella do you know the state of this, anything new?

@grigoryk I have no updates (tbh, I didn't think this would have much impact on fenix because it's a single module). I vaguely remember seeing pocmo revisiting this for ac several months ago on Slack but I don't know if anything came of that.

@pocmo Do you have any updates on parallel builds for fenix or ac?

**Comment by pocmo** [comment]
I do not really remember the specifics. Talked to @grigoryk recently and there were only two past incidents I remember, and I am not sure if any still apply or ever applied to Fenix:

* We saw flaky tests with `--parallel`.
* We had some tasks that did not correctly wait for other tasks they depend on when all ran in parallel (I think that was a problem with Glean once, but I think was fixed)

I think it's worth just testing it again, locally AND in automation - and gather new data. In one of the A-C issues I apparently saw it being slower on taskcluster when using `--parallel`. Not sure if that's still the case.
